core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -338,27 +352,20 @@ object TiBatchWrite {
     TiBatchWriteUtils.getRegionsByTable(tiContext.tiSession, table).toList
   }
 
-  private def extractHandleId(row: TiRow, tiTableInfo: TiTableInfo, isUpdate: Boolean): Long = {
+  private def extractHandleId(row: TiRow, tableInfo: TiTableInfo, isUpdate: Boolean): Long = {
     // If handle ID is changed when update, update will remove the old record first,
     // and then call `AddRecord` to add a new record.
     // Currently, only insert can set _tidb_rowid, update can not update _tidb_rowid.
-    val tblInfo = tiTableInfo.copyTableWithOutRowId()
-    val handle = if (row.fieldCount > tblInfo.getColumns.size && !isUpdate) {
+    val handle = if (row.fieldCount > tableInfo.getColumns.size && !isUpdate) {
       row.getLong(row.fieldCount - 1)
     } else {
-      val columnList = tiTableInfo.getColumns.asScala
+      val columnList = tableInfo.getColumns.asScala
       columnList.find(_.isPrimaryKey) match {
         case Some(primaryColumn) =>
           row.getLong(primaryColumn.getOffset)
```

Comment:
what if primary column type is not long?

Reply:
```
 val handle = if (row.fieldCount > tableInfo.getColumns.size && !isUpdate) {
      row.getLong(row.fieldCount - 1)
    }
```
covers this.

Reply:
if primary key exists, whether it's type is long or int, it is stored as `Long` in `com.pingcap.tikv.row.Row`?

Reply:
It stores the actual value in row.